aten/src/ATen/core/ATenDispatch.h

```diff
@@ -132,18 +135,97 @@ class CAFFE2_API ATenDispatch {
     return *this;
   }
 
+  ATenDispatch& registerFallbackBoxedOp(TensorTypeId id, FallbackBoxedFunction* fn) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    boxed_fallback_table_[static_cast<size_t>(id)] = fn;
+    return *this;
+  }
+
   const ATenOpTable* getOpTable(const char* schema) const {
     auto iter = op_tables_.find(schema);
     TORCH_CHECK(iter != op_tables_.end(),
         "No functions are registered for schema ", schema);
     return &iter->second;
   }
 
+  const FallbackBoxedFunction* getFallbackBoxedOp(TensorTypeId tid) const {
+    return boxed_fallback_table_[static_cast<size_t>(tid)];
+  }
+
  private:
   std::unordered_map<std::string, ATenOpTable> op_tables_;
+  FallbackBoxedFunction* boxed_fallback_table_[static_cast<int64_t>(TensorTypeId::NumTensorIds)] = {nullptr};
   std::mutex mutex_;
 };
 
 CAFFE2_API ATenDispatch& globalATenDispatch();
 
+template<
+  class Result, class... Args,
+  typename c10::guts::enable_if_t<
+    !supports_boxed_fallback<Result, Args...>::value,
+    std::nullptr_t
+  > = nullptr>
+Result callBoxedFallback(const char* schema, const FallbackBoxedFunction* boxed_fallback_fn, Args&&... args) {
+  // This is dead because we test the SFINAE condition before calling
+  // boxed_fallback_fn.  A more functional way of writing this with
+  // optional<Result> return value works poorly when void is involved.
+  TORCH_INTERNAL_ASSERT(0);
+}
+
+template<
+  class Result, class... Args,
+  typename c10::guts::enable_if_t<
+    supports_boxed_fallback<Result, Args...>::value,
+    std::nullptr_t
+  > = nullptr>
+Result callBoxedFallback(const char* schema, const FallbackBoxedFunction* boxed_fallback_fn, Args&&... args) {
+  torch::jit::Stack stack;
+  torch::jit::push(stack, std::forward<Args>(args)...);
+  boxed_fallback_fn(schema, &stack);
+  return torch::jit::pop(stack).to<Result>();
+}
+
+// NB: No universal forwarding
+template<class Result, class... Args>
+Result ATenOpTable::callUnboxed(Args... args) const {
+  using FuncType = Result(Args...);
+  // NB: No universal forwarding (takes const& only)
+  TensorTypeSet ts = detail::multi_dispatch_tensor_type_set(args...);
+  TensorTypeId tid = impl::dispatchTypeId(ts);
+
+  // You might think we can eliminate the second branch by maintaining a
+  // bitmask of registered operator keys, so we don't select dispatch ids
+  // which don't have implementations here.  But the net effect is that if you
+  // get a Variable CPUTensor, if there is no variable registration, you'll
+  // fall back to the CPU implementation.  Is this what you want?  Unlikely...
+
+  auto* unboxed_fn = reinterpret_cast<FuncType*>(function_table_[static_cast<int64_t>(tid)]);
+  if (C10_LIKELY(unboxed_fn != nullptr)) {
+    return (*unboxed_fn)(std::forward<Args>(args)...);
+  }
+
+  // We need to do a if statement on box fallback support AS WELL
+  // as SFINAE.
+  //    - If statement: because some functions return void, and so
```

Comment:
I find this explanation confusing. It sounds like the if statement and the SFINAE check different things, but they actually check exactly the same thing. It sounds like the if statement makes sure this works even if it's "return void", but that's actually not true. Since it's not an `if constexpr`, the compiler will still compile both branches, no matter the condition. The reason "return void" compiles fine is because that's an ok thing to do as long as the function itself returns void.